sbncode/SBNEventWeight/Calculators/BNBFlux/FluxCalcPrep.cxx

```diff
@@ -0,0 +1,352 @@
+              HadronAbriviation = "PP";
+            } else if(fprimaryHad[0] == -211){
+              HadronName = "PiMinus";
+              HadronAbriviation = "PM";
+            } else{ 
+              throw art::Exception(art::errors::StdException)
+                << "sanford-wang is only configured for charged pions ";
+            }
+            pname.push_back( Form("HARPData/%s/%sCrossSection",HadronName.c_str(),HadronAbriviation.c_str()) ); // Cross Section
+            pname.push_back( Form("HARPData/%s/%scovarianceMatrix",HadronName.c_str(),HadronAbriviation.c_str()) ); // Covariance Matrix
+            pname.push_back( Form("HARPData/%s/%smomentumBoundsArray",HadronName.c_str(),HadronAbriviation.c_str()) ); // Momentum Bounds
+            pname.push_back( Form("HARPData/%s/%sthetaBoundsArray",HadronName.c_str(),HadronAbriviation.c_str()) ); // Theta Bounds
+
+            HARPXSec = (TMatrixD*) file->Get(pname[0].c_str());
+            TMatrixD* HARPCov  = (TMatrixD*) file->Get(pname[1].c_str());
+
+            TDecompChol dc = TDecompChol(*(HARPCov));//perform Choleskey Decomposition
+            if(!dc.Decompose()){
+              throw art::Exception(art::errors::StdException)
+                << "Cannot decompose covariance matrix to begin smearing.";
+            }
+
+            //Get upper triangular matrix. This maintains the relations in the
+            //  covariance matrix, but simplifies the structure.
+            fIsDecomposed = true;
+            FitCov = new TMatrixD(dc.GetU());  
+
+
+            TArrayD* HARPmomentumBoundsArray = (TArrayD*) file->Get(pname[2].c_str());
+            HARPmomentumBounds = FluxWeightCalc::ConvertToVector(HARPmomentumBoundsArray);
+
+            TArrayD* HARPthetaBoundsArray = (TArrayD*) file->Get(pname[3].c_str());
+            HARPthetaBounds = FluxWeightCalc::ConvertToVector(HARPthetaBoundsArray);
+
+            /////////////////
+            //
+            //   Extract the Sanford-Wang Fit Parmeters
+            //
+            ////////////////
+
+            std::string ExternalFitInput = sp.find_file(fitInput);
+            TFile* Fitfile = new TFile(Form("%s",ExternalFitInput.c_str()));
+
+            std::string fitname; // these are what we will extract from the file
+            fitname = Form("SW/%s/SW%sFitVal",HadronName.c_str(),HadronName.c_str()); // Sanford-Wang Fit Parameters
+
+            TArrayD* SWParamArray = (TArrayD*) Fitfile->Get(fitname.c_str());
+            SWParam = FluxWeightCalc::ConvertToVector(SWParamArray);
+
+
+          }else  validC = false;//slightly incorrect calculator name in *fcl
+
+          if(validC){//load random numbers based on FitCov
+            //{2*multisim vector}
+            //{{Ncols() elements}} <-- feed these amount everytime;
+            for( int index = 1; index < 2*(FitCov->GetNcols()); index ++){
+              fParameterSet.Sample(engine);//load it 2*<number_of_multisims> times
+            }
```

Comment:
Why do you have to load this so many times? Maybe you could explain to me with this is doing specifically?

Reply:
Yes, this is to avoid segmentation fault. in case a random number causes a bad cross-section, i.e. negative value, then we use another (next) random number to estimate a new cross-section. We need sufficient random number to go through this process.

See the loop in the following section:
https://github.com/Ooohu/sbncode/blob/d324b83a33b9988b7091afd7624b4e0bb92b1e51/sbncode/SBNEventWeight/Calculators/BNBFlux/FluxCalcPrep.cxx#L290

This loop repeats until the size of `<vector> weights` is large enough, and `i++` always runs (line 290); In line 312, the size of `weights` grows only if a good `test_weight` is calculated, otherwise, the loop repeated with extra `i++`. Here `Vrandom[i]` at line 296 may go beyond the `Nuni` (number of multisims).

Reply:
OK, so `fParameterSet.Sample(engine);` generates a new random number, and stores in in a map as part of fParameterSet. Got it. Thanks